google-cloud-pubsub/lib/google/cloud/pubsub/publisher.rb

```diff
@@ -0,0 +1,195 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+require "monitor"
+require "concurrent"
+require "google/cloud/pubsub/publish_result"
+require "google/cloud/pubsub/service"
+
+module Google
+  module Cloud
+    module Pubsub
+      ##
+      # # Publisher
+      #
+      class Publisher
+        include MonitorMixin
+
+        MAX_COUNT = 2000
+        MAX_SIZE = 10000000
+
+        attr_reader :service, :queue
+        attr_reader :max_size, :max_count
+        attr_reader :thread_pool, :publish_task
+
+        def initialize service, max_size: MAX_SIZE, max_count: MAX_COUNT
+          @service = service
+          @max_size = max_size
+          @max_count = max_count
+
+          @queue = {}
+          @thread_pool = Concurrent::FixedThreadPool.new(
+            [2, Concurrent.processor_count * 2].max,
+            fallback_policy: :caller_runs
+          )
+          @publish_task = Concurrent::TimerTask.new(execution_interval: 0.25) do
+            flush
+          end
+
+          # init MonitorMixin
+          super()
+        end
+
+        def publish topic_name, data = nil, attributes = {}, &block
+          topic_name = service.topic_path topic_name
+          msg = create_pubsub_message data, attributes
+
+          synchronize do
+            batch = @queue[topic_name]
+            if batch.nil?
+              # Create a new batch.
+              batch = QueuedBatch.new msg, block
+            elsif !batch.try_add(msg, block, max_count: max_count,
+                                             max_size: max_size)
+              # We hit a threshold, publish existing batch.
+              publish_batch_async topic_name, batch
+              # Create a new batch.
+              batch = QueuedBatch.new msg, block
+            end
+            @queue[topic_name] = batch
+          end
+          nil
+        end
+
+        def start
+          @publish_task.execute
+        end
+
+        def stop
+          flush
+          @publish_task.shutdown
+        end
+
+        def started?
+          @publish_task.running?
+        end
+
+        def stopping?
+          @publish_task.shuttingdown?
+        end
+
+        def stopped?
+          @publish_task.shutdown?
+        end
+
+        def wait timeout = nil
+          @publish_task.wait_for_termination timeout
+        end
+
+        def flush
+          synchronize do
+            @queue.each do |topic_name, batch|
+              publish_batch_async topic_name, batch
+            end
+            @queue = {}
+          end
+        end
+
+        protected
+
+        def create_pubsub_message data, attributes
+          attributes ||= {}
+          if data.is_a?(::Hash) && attributes.empty?
```

Comment:
I am fine with a separate tracking issue and will not block on that question. Thanks!